web/api/octoprint_messages.py

```diff
@@ -48,46 +54,263 @@ def settings_dict(octoprint_settings):
+                pevent = PrintEvent.create(print, PrintEvent.DETECTED)
+
+            set_as_current = (
+                # when its unset or no changes for current print for a while
+                printer.current_print is None or
+                (
+                    (now - (printer.current_print.status_at or printer.current_print.updated_at)) >
+                    datetime.timedelta(seconds=CURRENT_SWITCH_OVER_SECS)
+                )
+            ) and (print.ended_at() is None)
+
+            if set_as_current:
+                printer.current_print = print
+                printer.save(update_fields=('updated_at', 'current_print'))
+
+                printer.printerprediction.reset_for_new_print()
+                send_should_watch_status = True
+
+    except Print.DoesNotExist:
+        raise ResurrectionError(
+            'Current print is deleted! '
+            'printer_id: {} | print_ts: {} | filename: {}'.format(
+                printer.id, ext_id, filename)
+        )
+
+    if send_should_watch_status:
+        printer.send_should_watch_status()
+
+    return (print, pevent)
+
+
+def process_print_event(printer: Printer,
+                        event_type: str,
+                        ext_id: int,
+                        filename: str,
+                        op_status: Dict,
+                        now: Optional[datetime.datetime] = None) -> Tuple[Print, Optional[PrintEvent], Optional[PrintEvent]]:
+    print: Optional[Print] = None
+    cevent: Optional[PrintEvent] = None
+    pevent: Optional[PrintEvent] = None
+    now = now or timezone.now()
+
+    if event_type == 'PrintStarted':
+        (print, pevent) = on_PrintStarted(printer, ext_id, filename, now)
+    else:
+        (print, cevent) = get_or_create_print(printer, ext_id, filename, now)
+
+        if event_type == 'PrintCancelled':
+            (print, pevent) = on_PrintCancelled(printer, print, now)
+        elif event_type == 'PrintFailed':
+            (print, pevent) = on_PrintFailed(printer, print, now)
+        elif event_type == 'PrintDone':
+            (print, pevent) = on_PrintDone(printer, print, now)
+        elif event_type == 'PrintLost':
+            (print, pevent) = on_PrintLost(printer, print, now)
+        elif event_type == 'PrintPaused':
+            (print, pevent) = on_PrintPaused(printer, print, now)
+        elif event_type == 'PrintResumed':
+            (print, pevent) = on_PrintResumed(printer, print, now)
+
+    # touching print and updating estimate
+    delta: datetime.timedelta = (now - (print.status_at or print.updated_at))
+    if print.ended_at() is None and delta > datetime.timedelta(PRINT_TOUCH_SECS):
+        print.status_at = now
+
+        next_estimate: Optional[datetime.datetime] = extract_estimate(op_status)
+        if next_estimate is not None:
+            print.estimated_finished_at = next_estimate
+
+        print.save(update_fields=('updated_at', 'status_at', 'estimated_finished_at'))
```

Comment:
Any way to make sure when in the future we change other field in the process, we will remember to add that field to this `update_fields` set?

Reply:
Something like this might help that. Interested?

+class UpdateWrapper:
+
+    def __init__(self, obj: models.Model, force: Optional[Tuple[str]]) -> None:
+        self._w_obj = obj
+        self._w_force = force or ()
+        self._w_changes: Dict[str, Any] = {}
+
+    def __setattr__(self, key: str, value: Any) -> None:
+        if key.startswith('_w_'):
+            self.__dict__[key] = value
+            return
+
+        self._w_changes[key] = value
+        setattr(self._w_obj, key, value)
+
+    def __getattr__(self, key: str) -> Any:
+        if key.startswith('_w_'):
+            return self.__dict__[key]
+        return getattr(self._w_obj, key)
+
+    def save(self, **kwargs) -> None:
+        keys = tuple(set(self._w_changes.keys()) | set(self._w_force))
+        self._w_obj.save(update_fields=keys, **kwargs)

Reply:
If the only reason to do it is to save the data transferred for the update SQL, I'd say we skip the `update_fields` param. The overhead for the update SQL should be small enough to be almost negligible. Don't feel it's worth the extra complexity and chances for error.

Reply:
I propose this to avoid data races. save() overwrites the row fully in the table and as these functions might alter only low number of fields this is how I would do it properly.

Reply:
An other rather basic option is to collect changes in a dict and apply them together. After playing a bit with the idea this updatewrapper still looks nice but might be missing some points, ruins type checks and adds some extra trouble when one needs to reference original object for setting related fields (thought I can simply wrap everything in the top function and forget it).